### PR TITLE
[Bug] Enable turn on/off libxsmm at runtime

### PR DESCRIPTION
**Pull request by peizhou001**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: src/array/context.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+ private:
+  bool _libxsmm = true;
+};
+
+DGL_ATEN_CONTEXT_EXPORT Context& globalContext();
+
+static inline Context& init() {
+    return globalContext();
+}
```
Comment (BarclayII) [inline src/array/context.h:45]: You don't need them.

File: src/array/cpu/spmm.h [hunk 2]
```diff
@@ -143,7 +144,8 @@ void SpMMSumCsr(const BcastOff& bcast, const CSRMatrix& csr, NDArray ufeat,
 #ifdef USE_LIBXSMM
   const bool no_libxsmm =
        bcast.use_bcast || std::is_same<DType, double>::value;
-  if (!no_libxsmm) {
+  const bool use_libxsmm = globalContext().libxsmm();
```
Comment (BarclayII) [inline src/array/cpu/spmm.h:147]: You can as well merge `use_libxsmm` into `no_libxsmm`.  I assume we don't need two variables.

File: src/array/kernel.cc [hunk 3]
```diff
@@ -490,6 +491,8 @@ std::pair<CSRMatrix, NDArray> CSRSum(
   return ret;
 }
 
+INIT_CONTEXT();
```
Comment (BarclayII) [inline src/array/kernel.cc:494]: You don't need this.

**Review by BarclayII — CHANGES_REQUESTED** [review]

File: src/array/cpu/spmm_blocking_libxsmm.h [hunk 4]
```diff
@@ -264,7 +264,8 @@ inline libxsmm_meltwfunction_opreduce_vecs_idx SpMMCreateLibxsmmKernel(
                (sizeof(IdType) == 8) ? LIBXSMM_DATATYPE_I64 : LIBXSMM_DATATYPE_I32, opredop_flags);
   }
   if (kernel == nullptr) {
-    LOG(FATAL) << "Failed to generate libxsmm kernel for the SpMM operation!";
+    LOG(FATAL) << "Failed to generate libxsmm kernel for the SpMM operation"
+                  "you can use API use_libxsmm(false) to diable it!";
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/cpu/spmm_blocking_libxsmm.h:268]: Fixed

File: src/array/context.cc [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file array/context.cc
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.cc:3]: Make sense, replace it with Config and leverage namespace to differentiate variety configs.

File: src/array/context.cc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+namespace aten {
+
+Context::Context() = default;
+
+Context& globalContext() {
+    static Context globalContext_;
+    return globalContext_;
+}
+
+void Context::setLibxsmm(bool b) {
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.cc:19]: Make sense, change the name to enableLibxsmm.

File: src/array/context.cc [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    _libxsmm = b;
+}
+
+bool Context::libxsmm() const {
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.cc:23]: Changed

File: src/array/context.h [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+#define DGL_ATEN_CONTEXT_EXPORT __declspec(dllexport)
+#else  // _WIN32
+#if defined(__GNUC__)
+#define DGL_ATEN_CONTEXT_EXPORT __attribute__((__visibility__("default")))
+#define DGL_ATEN_CONTEXT_HIDDEN __attribute__((__visibility__("hidden")))
+#else  // defined(__GNUC__)
+#define DGL_ATEN_CONTEXT_EXPORT
+#define DGL_ATEN_CONTEXT_HIDDEN
+#endif  // defined(__GNUC__)
+#endif  // _WIN32
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.h:25]: According to my investigation, we can use DGL_DLL as in GUN the default attribute is  export and for windows we always set "DDGL_EXPORTS" in compile flag. So use DGL_DLL here for concise

File: src/array/context.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+#define INIT_CONTEXT() static Context& context = init()
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.h:27]: Actually not, it just define the macro here, and the execution should happen where use it.

File: src/array/cpu/spmm.h (hunk 2 above)
Comment (peizhou001) [reply src/array/cpu/spmm.h:147]: Make sense, combine two variables.

File: src/array/context.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/context.h:27]: Oh you are right.  Only those who include the `INIT_CONTEXT` statement will have an instantiation.  I would still recommend using a static class method though.

File: src/array/context.cc [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace dgl {
+namespace aten {
+
+Context::Context() = default;
+
+Context& globalContext() {
+    static Context globalContext_;
+    return globalContext_;
+}
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.cc:17]: That' right, I originally followed the torch format, which use a global static object, but it seems the difference between it and a static in method is a global static can be initialized  before main method, but we do not need it.
Anyway, follow the tensordispatch way to use a method static for simplicity.

File: src/array/context.h (hunk 1 above)
Comment (peizhou001) [reply src/array/context.h:45]: Removed

File: src/array/kernel.cc (hunk 3 above)
Comment (peizhou001) [reply src/array/kernel.cc:494]: Removed

File: src/array/context.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (peizhou001) [reply src/array/context.h]: Make sense, and I have used static method for replacement.

File: src/array/config.h [hunk 11]
```diff
@@ -0,0 +1,34 @@
+/*!
```
Comment (BarclayII) [inline src/array/config.h:1]: Move this to `include/dgl/runtime`?  Other C++ global configs can be put here as well.

**Review by BarclayII — APPROVED** [review]
Minor comments.  Please feel free to merge it after resolving.  Thanks a lot!

File: src/array/config.h (hunk 11 above)
Comment (peizhou001) [reply src/array/config.h:1]: Sounds ok, and let me change the description from scope aten only to runtime.

File: include/dgl/runtime/config.h [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+class Config {
+ public:
+  static Config* Global() {
+    static Config config;
+    return &config;
+  }
+
+  // Enabling or disable use libxsmm for Spmm
+  void enableLibxsmm(bool);
```
Comment (jermainewang) [inline include/dgl/runtime/config.h:22]: Make sure you follow the Camel naming style: `EnableLibxsmm`, `IsLibxsmmAvailable` .

File: include/dgl/runtime/config.h [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ private:
+  Config() =  default;
+  bool _libxsmm = true;
+};
+
+}  // namespace runtime
+}  // namespace dgl
+
+#endif  // DGL_RUNTIME_CONFIG_H_
+
```
Comment (jermainewang) [inline include/dgl/runtime/config.h:34]: Remove extra newline

File: python/dgl/__init__.py [hunk 14]
```diff
@@ -50,6 +50,7 @@
 from . import optim
 from .frame import LazyFeature
 from .utils import apply_each
+from .sparse import is_libxsmm_enabled, use_libxsmm
```
Comment (jermainewang) [inline python/dgl/__init__.py:53]: It is strange to have these APIs in `sparse.py`. How about adding a `global.py` for all global configuration APIs?

File: python/dgl/sparse.py [hunk 15]
```diff
@@ -53,6 +55,26 @@ def infer_broadcast_shape(op, shp1, shp2):
     rst = tuple(max(d1, d2) for d1, d2 in zip(pad_shp1, pad_shp2))
     return rst[:-1] + (1,) if op == "dot" else rst
 
+def use_libxsmm(flag):
+    r"""Set whether Dgl use libxsmm for cpu Spmm operation at runtime.
+
+    Parameters
+    ----------
+    val : boolean
```
Comment (jermainewang) [inline python/dgl/sparse.py:63]: docstring mismatch `val` -> `flag`

File: python/dgl/sparse.py [hunk 16]
```diff
@@ -53,6 +55,26 @@ def infer_broadcast_shape(op, shp1, shp2):
     rst = tuple(max(d1, d2) for d1, d2 in zip(pad_shp1, pad_shp2))
     return rst[:-1] + (1,) if op == "dot" else rst
 
+def use_libxsmm(flag):
+    r"""Set whether Dgl use libxsmm for cpu Spmm operation at runtime.
```
Comment (jermainewang) [inline python/dgl/sparse.py:59]: * Use all capitalized `DGL`.
* Give a pointer to the libxsmm official repo or doc.
* `cpu` -> `CPU`
* `Spmm` needs explanation. To omit such burden, I suggest just say `Set whether DGL uses libxsmm at runtime.`

File: src/array/cpu/spmm_blocking_libxsmm.h [hunk 17]
```diff
@@ -264,7 +264,8 @@ inline libxsmm_meltwfunction_opreduce_vecs_idx SpMMCreateLibxsmmKernel(
                (sizeof(IdType) == 8) ? LIBXSMM_DATATYPE_I64 : LIBXSMM_DATATYPE_I32, opredop_flags);
   }
   if (kernel == nullptr) {
-    LOG(FATAL) << "Failed to generate libxsmm kernel for the SpMM operation!";
+    LOG(FATAL) << "Failed to generate libxsmm kernel for the SpMM operation."
+                  "You can use API dgl.use_libxsmm(false) to diable it!";
```
Comment (jermainewang) [inline src/array/cpu/spmm_blocking_libxsmm.h:268]:
```suggestion
                  " To disable libxsmm, use dgl.use_libxsmm(false).";
```
Be mindful of your tone. Try avoid second person in system messages. Be aware of spelling errors.

File: docs/source/api/python/dgl.rst [hunk 18]
```diff
@@ -208,7 +208,7 @@ set at each iteration. ``prop_edges_YYY`` applies traversal algorithm ``YYY`` an
 Utilities
 -----------------------------------------------
 
-Other utilities for controlling randomness, saving and loading graphs, functions that applies
+Other utilities for controlling randomness, saving and loading graphs, turn on/off libxsmm, functions that applies
```
Comment (jermainewang) [inline docs/source/api/python/dgl.rst:211]: To avoid future changes, use a more general term, e.g., `setting / getting runtime configurations`.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: include/dgl/runtime/config.h (hunk 12 above)
Comment (peizhou001) [reply include/dgl/runtime/config.h:22]: Fixed

File: include/dgl/runtime/config.h (hunk 13 above)
Comment (peizhou001) [reply include/dgl/runtime/config.h:34]: fixed

File: python/dgl/sparse.py (hunk 15 above)
Comment (peizhou001) [reply python/dgl/sparse.py:63]: fixed

File: python/dgl/sparse.py (hunk 16 above)
Comment (peizhou001) [reply python/dgl/sparse.py:59]: Make sense, use 'Set whether DGL uses libxsmm at runtime' instead and add the link to official repo of libxsmm.

File: src/array/cpu/spmm_blocking_libxsmm.h (hunk 17 above)
Comment (peizhou001) [reply src/array/cpu/spmm_blocking_libxsmm.h:268]: Fixed, I will be careful about these next time.

File: docs/source/api/python/dgl.rst (hunk 18 above)
Comment (peizhou001) [reply docs/source/api/python/dgl.rst:211]: Make sense, replace it with configuration description

File: python/dgl/__init__.py (hunk 14 above)
Comment (peizhou001) [reply python/dgl/__init__.py:53]: Make sense, as originally I just focused on spmm usage, let me add a "global.py" for future extension.

File: python/dgl/global_config.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+    Detailed information about libxsmm can be found here:
+    https://github.com/libxsmm/libxsmm
+
+    Parameters
+    ----------
+    flag : boolean
+        If True, use libxsmm, otherwise not.
+    """
+    SetLibxsmm(flag)
```
Comment (jermainewang) [inline python/dgl/global_config.py:18]: Our CAPI naming style is `_CAPI_XXX`. Pls change.

File: python/dgl/global_config.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+def is_libxsmm_enabled():
+    r"""Get whether the use_libxsmm flag is turned on.
+
+    Returns
+    ----------
+    use_libxsmm_flag[boolean]
+        True if the use_libxsmm flag is turned on.
```
Comment (jermainewang) [inline python/dgl/global_config.py:26]: You can add a see also section as:
```

See Also
----------
use_libxsmm
```

which will let users quickly locate highly relevant APIs

**Review by jermainewang — CHANGES_REQUESTED** [review]
The main part is fine but the PR lacks unit tests. Could you pls add?

File: python/dgl/global_config.py (hunk 20 above)
Comment (peizhou001) [reply python/dgl/global_config.py:26]: Got it, add it already.

File: python/dgl/global_config.py (hunk 19 above)
Comment (peizhou001) [reply python/dgl/global_config.py:18]: Fixed

**Review by jermainewang — APPROVED** [review]

**Comment by Rhett-Ying** [comment]
Commit ID: 66515ddbe01be2dc7cc44520a2cab2b4d6f78f8b

Build ID: 18

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/18/18/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/18/18/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: da014c435865b7f5f723bd7ebf24535e83c69f17

Build ID: 19

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/19/19/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/19/19/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: 21ccb68fe5e464940764cfe6f6e708d7ee59a1dd

Build ID: 20

Status: ❌ CI test failed in Stage [Distributed Torch CPU Unit test]. 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/20/20/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/20/20/logs/cireport.log)

**Comment by Rhett-Ying** [comment]
Commit ID: d2e45bfeb477437ef47f2658d509d2d3a7ba31fe

Build ID: 21

Status: ✅ CI test succeeded 

Report path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/21/21/logs/report.html) 

Full logs path: [link](https://dgl-ci-result.s3.us-west-2.amazonaws.com/dgl/PR-4455/21/21/logs/cireport.log)

**Pull request merged**